ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -0,0 +1,250 @@
+    // May need to encrypt buffer before upload
+    if (performUpload) {
+      BlobId blobId = (BlobId) messageInfo.getStoreKey();
+      cloudDestination.uploadBlob(blobId, size, new ByteBufferInputStream(messageBuf));
+    }
+  }
+
+  @Override
+  public void delete(MessageWriteSet messageSetToDelete) throws StoreException {
+    checkStarted();
+    checkDuplicates(messageSetToDelete);
+
+    try {
+      for (MessageInfo msgInfo : messageSetToDelete.getMessageSetInfo()) {
+        BlobId blobId = (BlobId) msgInfo.getStoreKey();
+        cloudDestination.deleteBlob(blobId);
+      }
+    } catch (Exception ex) {
+      throw new StoreException(ex, StoreErrorCodes.IOError);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * Currently, the only supported operation is to set the TTL to infinite (i.e. no arbitrary increase or decrease)
+   * @param messageSetToUpdate The list of messages that need to be updated
+   * @throws StoreException if there is a problem persisting the operation in the store.
+   */
+  @Override
+  public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException {
+    checkStarted();
+  }
+
+  @Override
+  public FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws StoreException {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public Set<StoreKey> findMissingKeys(List<StoreKey> keys) throws StoreException {
+    checkStarted();
+    // Check existence of keys in cloud container
+    Set<StoreKey> missingKeys = new HashSet<>();
+    try {
+      for (StoreKey key : keys) {
+        BlobId blobId = (BlobId) key;
+        if (cloudDestination.doesBlobExist(blobId) == false) {
+          missingKeys.add(key);
+        }
+      }
+    } catch (Exception ex) {
+      throw new StoreException(ex, StoreErrorCodes.IOError);
+    }
+    return missingKeys;
+  }
+
+  @Override
+  public StoreStats getStoreStats() {
+    return null;
+  }
+
+  @Override
+  public boolean isKeyDeleted(StoreKey key) throws StoreException {
+    checkStarted();
+    return false;
+  }
+
+  @Override
+  public long getSizeInBytes() {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return false;
+  }
+
+  @Override
+  public void shutdown() throws StoreException {
+    started = false;
+  }
+
+  /**
+   * @return {@code true} if this store has been started successfully.
+   */
+  boolean isStarted() {
+    return started;
+  }
+
+  private void checkStarted() throws StoreException {
+    if (!started) {
+      throw new StoreException("Store not started", StoreErrorCodes.Store_Not_Started);
+    }
+  }
+
+  /**
+   * Detects duplicates in {@code writeSet}
+   * @param writeSet the {@link MessageWriteSet} to detect duplicates in
+   * @throws StoreException if a duplicate is detected
+   */
+  // TODO: move into MessageWriteSet?
```

Comment:
Good idea. To avoid repeat code.

Reply:
Not easy to do since MessageWriteSet is an interface.  I removed the TODO